[BUG] Overflow in implicit constant conversion microseconds_from(uint64_t)

The function `sync::microseconds_from(...)` expects `long` (4 bytes) as an argument, but gets `uint64_t` (8 bytes). This casuses a warning.

buffer.cpp:1459:85: warning: overflow in implicit constant conversion  -Woverflow]:
```
m_tsTsbPdTimeBase += microseconds_from(uint64_t(CPacket::MAX_TIMESTAMP) + 1);
```

In this case `CPacket::MAX_TIMESTAMP = 0xFFFFFFFF` - an unsigned 32-bit value. So conversion to 4-bit long is wrong.

- SRT Version / commit ID: 1.4.1+ after PR #929.

## Possible resolution

Change the type of argument of the `sync::microseconds_from(...)` to `int64_t`

--- a/srtcore/buffer.cpp
+++ b/srtcore/buffer.cpp
@@ -1424,7 +1424,7 @@ steady_clock::time_point CRcvBuffer::getTsbPdTimeBase(uint32_t timestamp_us)
     * In this period, timestamps smaller than 30 seconds are considered to have wrapped around (then adjusted).
     * The wrap check period ends 30 seconds after the wrap point, afterwhich time base has been adjusted.
     */
-    uint64_t carryover = 0;
+    int64_t carryover = 0;
 
     // This function should generally return the timebase for the given timestamp_us.
     // It's assumed that the timestamp_us, for which this function is being called,
@@ -1448,7 +1448,7 @@ steady_clock::time_point CRcvBuffer::getTsbPdTimeBase(uint32_t timestamp_us)
 
         if (timestamp_us < TSBPD_WRAP_PERIOD)
         {
-            carryover = uint64_t(CPacket::MAX_TIMESTAMP) + 1;
+            carryover = int64_t(CPacket::MAX_TIMESTAMP) + 1;
         }
         //
         else if ((timestamp_us >= TSBPD_WRAP_PERIOD)
@@ -1456,7 +1456,7 @@ steady_clock::time_point CRcvBuffer::getTsbPdTimeBase(uint32_t timestamp_us)
         {
             /* Exiting wrap check period (if for packet delivery head) */
             m_bTsbPdWrapCheck = false;
-            m_tsTsbPdTimeBase += microseconds_from(uint64_t(CPacket::MAX_TIMESTAMP) + 1);
+            m_tsTsbPdTimeBase += microseconds_from(int64_t(CPacket::MAX_TIMESTAMP) + 1);
             tslog.Debug("tsbpd wrap period ends");
         }
     }

--- a/srtcore/buffer.cpp
+++ b/srtcore/buffer.cpp
@@ -1768,7 +1768,7 @@ void CRcvBuffer::readMsgHeavyLogging(int p)
     LOGC(dlog.Debug, log << CONID() << "readMsg: DELIVERED seq=" << seq
             << " T=" << FormatTime(srctime)
             << " in " << timediff_ms << "ms - TIME-PREVIOUS: PKT: "
-            << FormatTime(srctime) << " LOCAL: " << nowdiff_ms
+            << srctimediff_ms << " LOCAL: " << nowdiff_ms
             << " !" << BufferStamp(pkt.data(), pkt.size())
             << " NEXT pkt T=" << next_playtime);
 

--- a/srtcore/congctl.cpp
+++ b/srtcore/congctl.cpp
@@ -216,7 +216,7 @@ private:
         return SrtCongestion::SRM_FASTREXMIT;
     }
 
-    uint64_t updateNAKInterval(uint64_t nakint_us, int /*rcv_speed*/, size_t /*loss_length*/) ATR_OVERRIDE
+    int64_t updateNAKInterval(int64_t nakint_us, int /*rcv_speed*/, size_t /*loss_length*/) ATR_OVERRIDE
     {
         /*
          * duB:
@@ -237,7 +237,7 @@ private:
         return nakint_us / m_iNakReportAccel;
     }
 
-    uint64_t minNAKInterval() ATR_OVERRIDE
+    int64_t minNAKInterval() ATR_OVERRIDE
     {
         return m_iMinNakInterval_us;
     }

--- a/srtcore/congctl.h
+++ b/srtcore/congctl.h
@@ -192,15 +192,15 @@ public:
 
     virtual SrtCongestion::RexmitMethod rexmitMethod() = 0; // Implementation enforced.
 
-    virtual uint64_t updateNAKInterval(uint64_t nakint_us, int rcv_speed, size_t loss_length)
+    virtual int64_t updateNAKInterval(int64_t nakint_us, int rcv_speed, size_t loss_length)
     {
         if (rcv_speed > 0)
-            nakint_us += (loss_length * uint64_t(1000000) / rcv_speed);
+            nakint_us += (loss_length * int64_t(1000000) / rcv_speed);
 
         return nakint_us;
     }
 
-    virtual uint64_t minNAKInterval()
+    virtual int64_t minNAKInterval()
     {
         return 0; // Leave default
     }

--- a/srtcore/core.cpp
+++ b/srtcore/core.cpp
@@ -1274,16 +1274,16 @@ void CUDT::clearData()
     m_iDeliveryRate     = 16;
     m_iByteDeliveryRate = 16 * m_iMaxSRTPayloadSize;
     m_iAckSeqNo         = 0;
-    m_tsLastAckTime       = srt::sync::steady_clock::now();
+    m_tsLastAckTime     = steady_clock::now();
 
     // trace information
     {
         CGuard stat_lock(m_StatsLock);
 
-        m_stats.tsStartTime = srt::sync::steady_clock::now();
+        m_stats.tsStartTime = steady_clock::now();
         m_stats.sentTotal = m_stats.recvTotal = m_stats.sndLossTotal = m_stats.rcvLossTotal = m_stats.retransTotal =
             m_stats.sentACKTotal = m_stats.recvACKTotal = m_stats.sentNAKTotal = m_stats.recvNAKTotal = 0;
-        m_stats.tsLastSampleTime = srt::sync::steady_clock::now();
+        m_stats.tsLastSampleTime = steady_clock::now();
         m_stats.traceSent = m_stats.traceRecv = m_stats.traceSndLoss = m_stats.traceRcvLoss = m_stats.traceRetrans =
             m_stats.sentACK = m_stats.recvACK = m_stats.sentNAK = m_stats.recvNAK = 0;
         m_stats.traceRcvRetrans                                                   = 0;
@@ -1337,7 +1337,7 @@ void CUDT::clearData()
     m_bPeerRexmitFlag = false;
 
     m_RdvState         = CHandShake::RDV_INVALID;
-    m_tsRcvPeerStartTime = srt::sync::steady_clock::time_point();
+    m_tsRcvPeerStartTime = steady_clock::time_point();
 }
 
 void CUDT::open()
@@ -1507,7 +1507,7 @@ size_t CUDT::fillSrtHandshake_HSRSP(uint32_t *srtdata, size_t /* srtlen - unused
     // Setting m_ullRcvPeerStartTime is done in processSrtMsg_HSREQ(), so
     // this condition will be skipped only if this function is called without
     // getting first received HSREQ. Doesn't look possible in both HSv4 and HSv5.
-    if (srt::sync::is_zero(m_tsRcvPeerStartTime))
+    if (is_zero(m_tsRcvPeerStartTime))
     {
         LOGC(mglog.Fatal, log << "IPE: fillSrtHandshake_HSRSP: m_tsRcvPeerStartTime NOT SET!");
         return 0;
@@ -1733,7 +1733,7 @@ bool CUDT::createSrtHandshake(ref_t<CPacket>    r_pkt,
         // If fillSrtHandshake_HSRSP catches the condition of m_tsRcvPeerStartTime == 0,
         // it will return size 0, which will mess up with further extension procedures;
         // PREVENT THIS HERE.
-        if (hs.m_iReqType == URQ_CONCLUSION && srths_cmd == SRT_CMD_HSRSP && srt::sync::is_zero(m_tsRcvPeerStartTime))
+        if (hs.m_iReqType == URQ_CONCLUSION && srths_cmd == SRT_CMD_HSRSP && is_zero(m_tsRcvPeerStartTime))
         {
             LOGC(mglog.Error,
                  log << "createSrtHandshake: IPE (non-fatal): Attempting to craft HSRSP without received HSREQ. "
@@ -5179,8 +5179,8 @@ SRT_REJECT_REASON CUDT::setupCC()
     // Override the value of minimum NAK interval, per SrtCongestion's wish.
     // When default 0 value is returned, the current value set by CUDT
     // is preserved.
-    const srt::sync::steady_clock::duration min_nak = srt::sync::microseconds_from(m_CongCtl->minNAKInterval());
-    if (min_nak != srt::sync::steady_clock::duration::zero())
+    const steady_clock::duration min_nak = microseconds_from(m_CongCtl->minNAKInterval());
+    if (min_nak != steady_clock::duration::zero())
         m_tdMinNakInterval = min_nak;
 
     // Update timers
@@ -5710,7 +5710,7 @@ int CUDT::sendmsg2(const char *data, int len, ref_t<SRT_MSGCTRL> r_mctrl)
     {
         // delay the EXP timer to avoid mis-fired timeout
         CGuard ack_lock(m_RecvAckLock);
-        m_tsLastRspAckTime = srt::sync::steady_clock::now();
+        m_tsLastRspAckTime = steady_clock::now();
         m_iReXmitCount   = 1;
     }
 
@@ -6088,7 +6088,7 @@ int64_t CUDT::sendfile(fstream &ifs, int64_t &offset, int64_t size, int block)
     if (m_pSndBuffer->getCurrBufSize() == 0)
     {
         // delay the EXP timer to avoid mis-fired timeout
-        m_tsLastRspAckTime = srt::sync::steady_clock::now();
+        m_tsLastRspAckTime = steady_clock::now();
         m_iReXmitCount   = 1;
     }
 
@@ -6613,7 +6613,7 @@ bool CUDT::updateCC(ETransmissionEvent evt, EventVariant arg)
         // NOTE: THESE things come from CCC class:
         // - m_dPktSndPeriod
         // - m_dCWndSize
-        m_tdSendInterval      = srt::sync::microseconds_from((int64_t)m_CongCtl->pktSndPeriod_us());
+        m_tdSendInterval    = microseconds_from((int64_t)m_CongCtl->pktSndPeriod_us());
         m_dCongestionWindow = m_CongCtl->cgWindowSize();
 #if ENABLE_HEAVY_LOGGING
         HLOGC(mglog.Debug,
@@ -6827,8 +6827,8 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const void *lparam, void *rparam, in
         else if (ack == m_iRcvLastAck)
         {
             // If the ACK was just sent already AND elapsed time did not exceed RTT,
-            if ((srt::sync::steady_clock::now() - m_tsLastAckTime) <
-                (srt::sync::microseconds_from(m_iRTT + 4 * m_iRTTVar)))
+            if ((steady_clock::now() - m_tsLastAckTime) <
+                (microseconds_from(m_iRTT + 4 * m_iRTTVar)))
             {
                 CGuard::leaveCS(m_RcvBufferLock);
                 break;
@@ -6864,7 +6864,7 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const void *lparam, void *rparam, in
             if (data[ACKD_BUFFERLEFT] < 2)
                 data[ACKD_BUFFERLEFT] = 2;
 
-            if (srt::sync::steady_clock::now() - m_tsLastAckTime > m_tdACKInterval)
+            if (steady_clock::now() - m_tsLastAckTime > m_tdACKInterval)
             {
                 int rcvRate;
                 int ctrlsz = ACKD_TOTAL_SIZE_UDTBASE * ACKD_FIELD_SIZE; // Minimum required size
@@ -6888,7 +6888,7 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const void *lparam, void *rparam, in
                 // ELSE: leave the buffer with ...UDTBASE size.
 
                 ctrlpkt.pack(pkttype, &m_iAckSeqNo, data, ctrlsz);
-                m_tsLastAckTime = srt::sync::steady_clock::now();
+                m_tsLastAckTime = steady_clock::now();
             }
             else
             {
@@ -6962,11 +6962,11 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const void *lparam, void *rparam, in
         }
 
         // update next NAK time, which should wait enough time for the retansmission, but not too long
-        m_tdNAKInterval = srt::sync::microseconds_from(m_iRTT + 4 * m_iRTTVar);
+        m_tdNAKInterval = microseconds_from(m_iRTT + 4 * m_iRTTVar);
 
         // Fix the NAKreport period according to the congctl
         m_tdNAKInterval =
-            srt::sync::microseconds_from(m_CongCtl->updateNAKInterval(srt::sync::count_microseconds(m_tdNAKInterval),
+            microseconds_from(m_CongCtl->updateNAKInterval(count_microseconds(m_tdNAKInterval),
                                                                       m_RcvTimeWindow.getPktRcvSpeed(),
                                                                       m_pRcvLossList->getLossLength()));
 
@@ -7077,7 +7077,7 @@ void CUDT::updateSndLossListOnACK(int32_t ackdata_seqno)
     CGuard::leaveCS(m_StatsLock);
 }
 
-void CUDT::processCtrlAck(const CPacket &ctrlpkt, const srt::sync::steady_clock::time_point& currtime)
+void CUDT::processCtrlAck(const CPacket &ctrlpkt, const steady_clock::time_point& currtime)
 {
     const int32_t *ackdata       = (const int32_t *)ctrlpkt.m_pcData;
     const int32_t  ackdata_seqno = ackdata[ACKD_RCVLASTACK];
@@ -7541,7 +7541,7 @@ void CUDT::processCtrl(CPacket &ctrlpkt)
                 const int nbsent      = m_pSndQueue->sendto(m_pPeerAddr, response);
                 if (nbsent)
                 {
-                    m_tsLastSndTime = srt::sync::steady_clock::now();
+                    m_tsLastSndTime = steady_clock::now();
                 }
             }
         }
@@ -8043,7 +8043,7 @@ int CUDT::processData(CUnit *in_unit)
    // m_pRcvBuffer->addLocalTsbPdDriftSample(packet.getMsgTimeStamp());
    // Just heard from the peer, reset the expiration count.
    m_iEXPCount = 1;
-   m_tsLastRspTime = srt::sync::steady_clock::now();
+   m_tsLastRspTime = steady_clock::now();
 
     // We are receiving data, start tsbpd thread if TsbPd is enabled
     if (m_bTsbPd && pthread_equal(m_RcvTsbPdThread, pthread_t()))
@@ -9090,8 +9090,8 @@ void CUDT::checkACKTimer(const steady_clock::time_point &currtime)
         // ACK timer expired or ACK interval is reached
         sendCtrl(UMSG_ACK);
 
-        const srt::sync::steady_clock::duration ack_interval = m_CongCtl->ACKTimeout_us() > 0
-            ? srt::sync::microseconds_from(m_CongCtl->ACKTimeout_us())
+        const steady_clock::duration ack_interval = m_CongCtl->ACKTimeout_us() > 0
+            ? microseconds_from(m_CongCtl->ACKTimeout_us())
             : m_tdACKInterval;
         m_tsNextACKTime = currtime + ack_interval;
 
@@ -9156,12 +9156,12 @@ bool CUDT::checkExpTimer(const steady_clock::time_point& currtime)
     steady_clock::time_point next_exp_time;
     if (m_CongCtl->RTO())
     {
-        next_exp_time = m_tsLastRspTime + srt::sync::microseconds_from(m_CongCtl->RTO());
+        next_exp_time = m_tsLastRspTime + microseconds_from(m_CongCtl->RTO());
     }
     else
     {
-        srt::sync::steady_clock::duration exp_timeout =
-            srt::sync::microseconds_from(m_iEXPCount * (m_iRTT + 4 * m_iRTTVar) + COMM_SYN_INTERVAL_US);
+        steady_clock::duration exp_timeout =
+            microseconds_from(m_iEXPCount * (m_iRTT + 4 * m_iRTTVar) + COMM_SYN_INTERVAL_US);
         if (exp_timeout < (m_iEXPCount * m_tdMinExpInterval))
             exp_timeout = m_iEXPCount * m_tdMinExpInterval;
         next_exp_time = m_tsLastRspTime + exp_timeout;
@@ -9175,7 +9175,7 @@ bool CUDT::checkExpTimer(const steady_clock::time_point& currtime)
     // Haven't received any information from the peer, is it dead?!
     // timeout: at least 16 expirations and must be greater than 5 seconds
     if ((m_iEXPCount > COMM_RESPONSE_MAX_EXP) &&
-        (currtime - m_tsLastRspTime > srt::sync::microseconds_from(PEER_IDLE_TMO_US)))
+        (currtime - m_tsLastRspTime > microseconds_from(PEER_IDLE_TMO_US)))
     {
         //
         // Connection is broken.
@@ -9203,7 +9203,7 @@ bool CUDT::checkExpTimer(const steady_clock::time_point& currtime)
 
     HLOGC(mglog.Debug,
           log << "EXP TIMER: count=" << m_iEXPCount << "/" << (+COMM_RESPONSE_MAX_EXP) << " elapsed="
-              << (srt::sync::count_microseconds(currtime - m_tsLastRspTime)) << "/" << (+PEER_IDLE_TMO_US) << "us");
+              << (count_microseconds(currtime - m_tsLastRspTime)) << "/" << (+PEER_IDLE_TMO_US) << "us");
 
     ++m_iEXPCount;
 
@@ -9244,7 +9244,7 @@ void CUDT::checkRexmitTimer(const steady_clock::time_point& currtime)
     const uint64_t rtt_syn = (m_iRTT + 4 * m_iRTTVar + 2 * COMM_SYN_INTERVAL_US);
     const uint64_t exp_int_us = (m_iReXmitCount * rtt_syn + COMM_SYN_INTERVAL_US);
 
-    if (currtime <= (m_tsLastRspAckTime + srt::sync::microseconds_from(exp_int_us)))
+    if (currtime <= (m_tsLastRspAckTime + microseconds_from(exp_int_us)))
         return;
 
     // If there is no unacknowledged data in the sending buffer,
@@ -9328,7 +9328,7 @@ void CUDT::checkTimers()
     // Check if FAST or LATE packet retransmission is required
     checkRexmitTimer(currtime);
 
-    if (currtime > m_tsLastSndTime + srt::sync::microseconds_from(COMM_KEEPALIVE_PERIOD_US))
+    if (currtime > m_tsLastSndTime + microseconds_from(COMM_KEEPALIVE_PERIOD_US))
     {
         sendCtrl(UMSG_KEEPALIVE);
         HLOGP(mglog.Debug, "KEEPALIVE");

--- a/srtcore/core.cpp
+++ b/srtcore/core.cpp
@@ -6613,7 +6613,7 @@ bool CUDT::updateCC(ETransmissionEvent evt, EventVariant arg)
         // NOTE: THESE things come from CCC class:
         // - m_dPktSndPeriod
         // - m_dCWndSize
-        m_tdSendInterval      = srt::sync::microseconds_from((long)m_CongCtl->pktSndPeriod_us());
+        m_tdSendInterval      = srt::sync::microseconds_from((int64_t)m_CongCtl->pktSndPeriod_us());
         m_dCongestionWindow = m_CongCtl->cgWindowSize();
 #if ENABLE_HEAVY_LOGGING
         HLOGC(mglog.Debug,

--- a/srtcore/sync.cpp
+++ b/srtcore/sync.cpp
@@ -1,4 +1,14 @@
-﻿#include <iomanip>
+﻿/*
+ * SRT - Secure, Reliable, Transport
+ * Copyright (c) 2019 Haivision Systems Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ */
+
+#include <iomanip>
 #include <math.h>
 #include <stdexcept>
 #include "sync.h"

--- a/srtcore/sync.cpp
+++ b/srtcore/sync.cpp
@@ -118,32 +118,32 @@ srt::sync::TimePoint<srt::sync::steady_clock> srt::sync::steady_clock::now()
     return TimePoint<steady_clock>(x);
 }
 
-long long srt::sync::count_microseconds(const steady_clock::duration& t)
+int64_t srt::sync::count_microseconds(const steady_clock::duration& t)
 {
     return t.count() / s_cpu_frequency;
 }
 
-long long srt::sync::count_milliseconds(const steady_clock::duration& t)
+int64_t srt::sync::count_milliseconds(const steady_clock::duration& t)
 {
     return t.count() / s_cpu_frequency / 1000;
 }
 
-long long srt::sync::count_seconds(const steady_clock::duration& t)
+int64_t srt::sync::count_seconds(const steady_clock::duration& t)
 {
     return t.count() / s_cpu_frequency / 1000000;
 }
 
-srt::sync::steady_clock::duration srt::sync::microseconds_from(long t_us)
+srt::sync::steady_clock::duration srt::sync::microseconds_from(int64_t t_us)
 {
     return steady_clock::duration(t_us * s_cpu_frequency);
 }
 
-srt::sync::steady_clock::duration srt::sync::milliseconds_from(long t_ms)
+srt::sync::steady_clock::duration srt::sync::milliseconds_from(int64_t t_ms)
 {
     return steady_clock::duration((1000 * t_ms) * s_cpu_frequency);
 }
 
-srt::sync::steady_clock::duration srt::sync::seconds_from(long t_s)
+srt::sync::steady_clock::duration srt::sync::seconds_from(int64_t t_s)
 {
     return steady_clock::duration((1000000 * t_s) * s_cpu_frequency);
 }

--- a/srtcore/sync.h
+++ b/srtcore/sync.h
@@ -1,6 +1,6 @@
 /*
  * SRT - Secure, Reliable, Transport
- * Copyright (c) 2018 Haivision Systems Inc.
+ * Copyright (c) 2019 Haivision Systems Inc.
  *
  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
@@ -8,6 +8,8 @@
  *
  */
 #pragma once
+#ifndef __SRT_SYNC_H__
+#define __SRT_SYNC_H__
 
 //#define USE_STL_CHRONO
 //#define ENABLE_CXX17
@@ -209,3 +211,5 @@ std::string FormatTimeSys(const steady_clock::time_point& time);
 
 }; // namespace sync
 }; // namespace srt
+
+#endif // __SRT_SYNC_H__

--- a/srtcore/sync.h
+++ b/srtcore/sync.h
@@ -157,18 +157,18 @@ inline Duration<steady_clock> operator*(const int& lhs, const Duration<steady_cl
     return rhs * lhs;
 }
 
-inline long long count_microseconds(const TimePoint<steady_clock> tp)
+inline int64_t count_microseconds(const TimePoint<steady_clock> tp)
 {
-    return static_cast<long long>(tp.us_since_epoch());
+    return static_cast<int64_t>(tp.us_since_epoch());
 }
 
-long long count_microseconds(const steady_clock::duration& t);
-long long count_milliseconds(const steady_clock::duration& t);
-long long count_seconds(const steady_clock::duration& t);
+int64_t count_microseconds(const steady_clock::duration& t);
+int64_t count_milliseconds(const steady_clock::duration& t);
+int64_t count_seconds(const steady_clock::duration& t);
 
-Duration<steady_clock> microseconds_from(long t_us);
-Duration<steady_clock> milliseconds_from(long t_ms);
-Duration<steady_clock> seconds_from(long t_s);
+Duration<steady_clock> microseconds_from(int64_t t_us);
+Duration<steady_clock> milliseconds_from(int64_t t_ms);
+Duration<steady_clock> seconds_from(int64_t t_s);
 
 inline bool is_zero(const TimePoint<steady_clock>& t) { return t.is_zero(); }
 

--- a/test/test_sync.cpp
+++ b/test/test_sync.cpp
@@ -46,6 +46,21 @@ TEST(SyncDuration, BasicChecks)
     EXPECT_EQ(count_seconds(a), 0);
 }
 
+/// Check operations on (uint32_t + 1)
+TEST(SyncDuration, DurationFrom)
+{
+    const int64_t val = int64_t(numeric_limits<uint32_t>::max()) + 1;
+
+    const steady_clock::duration us_from = microseconds_from(val);
+    EXPECT_EQ(count_microseconds(us_from), val);
+
+    const steady_clock::duration ms_from = milliseconds_from(val);
+    EXPECT_EQ(count_milliseconds(ms_from), val);
+
+    const steady_clock::duration s_from = seconds_from(val);
+    EXPECT_EQ(count_seconds(s_from), val);
+}
+
 TEST(SyncDuration, RelOperators)
 {
     const steady_clock::duration a;

